src/git/commit.rs

```diff
@@ -34,6 +34,72 @@ impl Repository {
             Err(Git2Error::NothingToCommit { branch, statuses })
         }
     }
+
+    fn commit_or_signed_commit(
+        &self,
+        sig: &Signature,
+        commit_message: &str,
+        tree: &Tree,
+        parents: &[&Commit],
+        sign: bool,
+    ) -> Result<Oid, Git2Error> {
+        if sign {
+            let commit_buf =
+                self.0
+                    .commit_create_buffer(sig, sig, commit_message, tree, parents)?;
+
+            let commit_as_str = std::str::from_utf8(&commit_buf)
+                .expect("Invalid UTF-8 commit message")
+                .to_string();
+
+            let key = self.signin_key().ok();
+            let gpg_signature = gpg_sign_string(key, &commit_as_str)?;
+            let oid = self
+                .0
+                .commit_signed(&commit_as_str, &gpg_signature, Some("gpgsig"))?;
+
+            // This is needed because git2 does not update HEAD after creating a signed commit
+            let commit = self.0.find_object(oid, Some(ObjectType::Commit))?;
+            self.0.reset(&commit, ResetType::Mixed, None)?;
+            Ok(oid)
+        } else {
+            self.0
+                .commit(Some("HEAD"), sig, sig, commit_message, tree, parents)
+        }
+        .map_err(Git2Error::Other)
+    }
+}
+
+fn gpg_sign_string(key: Option<String>, content: &str) -> Result<String, Git2Error> {
+    let mut child = Command::new("gpg");
+    child.arg("--armor");
+    child.arg("--detach-sig");
+
+    if let Some(key) = key {
+        child.arg("--default-key");
+        child.arg(key);
```

Comment:
maybe `child.args(["--default-key", key])`, then we'd have to autoref in if let: `if let Some(key) = &key`